Enriched egress events from flyteadmin

pingsutw opened the pull request:
# TL;DR
Allow flyteadmin to publish cloud events to the new topic

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
execution_manager publish a cloudevent to the new topic when invoking [CreateWorkflowEvent](https://github.com/flyteorg/flyteadmin/blob/412b834a76c98edfe95bc82eda6a7d23da504c9a/pkg/manager/impl/execution_manager.go#L1162)

## Tracking Issue
https://github.com/flyteorg/flyte/issues/1019

## Follow-up issue
_NA_

wild-endeavor commented on pkg/async/notifications/implementations/cloudevent_publisher.go:
```diff
@@ -0,0 +1,94 @@
+)
+
+const (
+	cloudEventType   = "com.flyte.workflow"
```
i think for the initial implementation we only need one or two events, but I think the structure should be there to be able to translate all events we have - all of these https://github.com/flyteorg/flyteidl/blob/master/protos/flyteidl/event/event.proto

Where is the filter currently to decide which events to translate from raw event to cloudevent?

pingsutw replied:
> Where is the filter currently to decide which events to translate from raw event to cloudevent?

here. https://github.com/flyteorg/flyteadmin/pull/355/files#diff-899da2dc4f47b8d6467d0d2a8d779f36fcb7334965153ceae018e2652e9f0ac7R36-R38

wild-endeavor reviewed (commented):
I haven't looked too closely at the CloudEvent spec.  It doesn't feel like it does a whole lot right now.  It feels like we're just doing event.SetData... which is fine but it doesn't really buy us that much.  I was hoping there would at least be a json schema or something we could publish as part of the cloudevent spec.

does this make sense?  the broader goal of going through cloudevent was that we could transform the raw flyte events into a standardized set of even information that wouldn't change even if the backend raw events changed.

pingsutw commented on Dockerfile:
```diff
@@ -11,14 +11,14 @@ COPY go.mod go.sum /go/src/github.com/flyteorg/flyteadmin/
 
 # Pull dependencies
-RUN go mod download
+RUN go mod vendor && go mod download
```
The error occurred every time I built the image `make docker_build`, so I added it. It only happened when I changed the code

anyway, will remove it before merging.

```bash
#13 0.249 	k8s.io/utils@v0.0.0-20210305010621-2afb4311ab10: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
#13 0.249 	sigs.k8s.io/structured-merge-diff/v4@v4.1.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
#13 0.249 	sigs.k8s.io/yaml@v1.2.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
#13 0.249 	github.com/robfig/cron/v3: is replaced in go.mod, but not marked as replaced in vendor/modules.txt
#13 0.249 
#13 0.249 	To ignore the vendor directory, use -mod=readonly or -mod=mod.
#13 0.249 	To sync the vendor directory, run:
#13 0.249 		go mod vendor
#13 0.250 make: *** [Makefile:46: linux_compile] Error 1
------
executor failed running [/bin/sh -c make linux_compile]: exit code: 2
make: *** [docker_build] Error 1
```

wild-endeavor replied:
that's weird.  @EngHabu do you have any ideas?

katrogan commented on pkg/async/notifications/implementations/cloudevent_publisher.go:
```diff
@@ -0,0 +1,133 @@
+	event.SetSource(cloudEventSource)
+	event.SetID(uuid.New().String())
+	event.SetTime(time.Now())
+	event.SetExtension(jsonSchemaURL, "https://github.com/pingsutw/flyteidl/blob/cloudevent2/jsonschema/workflow_execution.json")
```
is this final? are we checking this into idl?

pingsutw replied:
sorry, the link is outdated. The new schema url is https://github.com/flyteorg/flyteidl/blob/cloudevent2/jsonschema/workflow_execution.json

katrogan commented on pkg/async/notifications/implementations/cloudevent_publisher.go:
```diff
@@ -0,0 +1,133 @@
+		if e, found := supportedEvents[event]; found {
+			eventSet = eventSet.Insert(e)
+		} else {
+			logger.Errorf(context.Background(), "Unsupported event type [%s] in the config")
```
panic here?

katrogan reviewed (commented):
top-level questions

- do we want calls to publish need to be async so that they don't block the execution event write path? is lossiness okay?
- for clarity, instead of adding this code in notifications/ can we create a separate folder under async/

pingsutw commented on pkg/async/notifications/implementations/cloudevent_publisher.go:
```diff
@@ -0,0 +1,133 @@
+}
+
+func (p *CloudEventPublisher) Publish(ctx context.Context, notificationType string, msg proto.Message) error {
+	p.systemMetrics.PublishTotal.Inc()
```
okay.
Should we also move it below the call to `!p.shouldPublishEvent`  in `event_publisher`?
https://github.com/flyteorg/flyteadmin/blob/c80c8d8fd97f607da83db3ebb84cf0aec3b1f191/pkg/async/notifications/implementations/event_publisher.go#L53

pingsutw commented:
> do we want calls to publish need to be async so that they don't block the execution event write path? is lossiness okay?

I've updated the code. To minimize overhead, use goroutine to publish cloudevent instead.

> for clarity, instead of adding this code in notifications/ can we create a separate folder under async/

Updated it.

katrogan commented on pkg/async/cloudevent/cloudevent_publisher.go:
```diff
@@ -0,0 +1,142 @@
+func (s *KafkaSender) Send(ctx context.Context, notificationType string, event cloudevents.Event) error {
+	if result := s.Client.Send(
+		// Set the producer message key
+		kafka_sarama.WithMessageKey(context.Background(), sarama.StringEncoder(event.ID())),
```
why not re-use the input context?

pingsutw replied:
updated it, sorry. my bad

katrogan commented on pkg/async/cloudevent/cloudevent_publisher.go:
```diff
@@ -0,0 +1,142 @@
+
+	if err := event.SetData(cloudevents.ApplicationJSON, &msg); err != nil {
+		p.systemMetrics.PublishError.Inc()
+		logger.Errorf(ctx, "Failed to encode data with error: %v", err)
```
nit: include the original msg here too?

katrogan commented on pkg/async/notifications/factory.go:
```diff
@@ -229,3 +236,67 @@ func NewEventsPublisher(config runtimeInterfaces.ExternalEventsConfig, scope pro
 	}
 }
+
+func NewCloudEventsPublisher(config runtimeInterfaces.CloudEventsConfig, scope promutils.Scope) interfaces.Publisher {
```
can we move this factory method under the cloudevent dir too?

katrogan commented on pkg/async/notifications/factory.go:
```diff
@@ -229,3 +236,67 @@ func NewEventsPublisher(config runtimeInterfaces.ExternalEventsConfig, scope pro
+		if err != nil {
+			panic(err)
+		}
+		defer sender.Close(context.Background())
```
let's pass in a context into this func and use that here

katrogan commented on pkg/runtime/application_config_provider.go:
```diff
@@ -86,6 +87,10 @@ var externalEventsConfig = config.MustRegisterSection(externalEvents, &interface
 })
 
+var cloudEventsConfig = config.MustRegisterSection(cloudEvents, &interfaces.CloudEventsConfig{
+	Type: common.Local,
```
can we add sensible defaults for `ReconnectAttempts` and `ReconnectDelaySeconds`?

katrogan replied:
ah okay, staying consistent is fine too

katrogan commented on pkg/runtime/interfaces/application_configuration.go:
```diff
@@ -433,6 +438,24 @@ type ExternalEventsConfig struct {
+	GCPConfig   GCPConfig   `json:"gcp"`
+	KafkaConfig KafkaConfig `json:"kafka"`
+	// kafka broker addresses
+	Brokers []string `json:"brokers"`
```
should this move to the KafkaConfig?

katrogan reviewed (commented):
this is looking great! thanks for adding the comprehensive unit tests

pingsutw commented on pkg/async/cloudevent/cloudevent_publisher.go:
```diff
@@ -0,0 +1,142 @@
+	event.SetType(notificationType)
+	event.SetSource(cloudEventSource)
+	event.SetID(uuid.New().String())
+	event.SetTime(time.Now())
```
Updated it

pingsutw commented on pkg/async/cloudevent/implementations/cloudevent_publisher_test.go:
```diff
@@ -0,0 +1,200 @@
+	"github.com/stretchr/testify/assert"
+)
+
+type mockKafKaSender struct{}
```
sorry, typo

pingsutw commented on pkg/async/cloudevent/implementations/sender.go:
```diff
@@ -0,0 +1,60 @@
+		kafka_sarama.WithMessageKey(ctx, sarama.StringEncoder(event.ID())),
+		event,
+	); cloudevents.IsUndelivered(result) {
+		return fmt.Errorf("failed to send cloud event: %v", result)
```
I think we don't need it. There is a retry mechanism in cloudevent SDK, so we don't need to handle it.

katrogan replied:
awesome

pingsutw commented on pkg/async/cloudevent/implementations/sender.go:
```diff
@@ -0,0 +1,60 @@
+)
+
+// Sender Defines the interface for sending cloudevents.
+type Sender interface {
```
updated it

pingsutw replied:
The new file structure will be like
```
cloudevent
├── factory.go
├── factory_test.go
├── implementations
│   ├── cloudevent_publisher.go
│   ├── cloudevent_publisher_test.go
│   ├── sender.go
│   └── sender_test.go
├── interfaces
│   ├── publisher.go
│   └── sender.go
└── mocks
    ├── publisher.go
    └── sender.go
```

katrogan replied:
great, thank you!

pingsutw commented on pkg/async/cloudevent/factory.go:
```diff
@@ -0,0 +1,85 @@
+	"github.com/flyteorg/flytestdlib/promutils"
+)
+
+func NewCloudEventsPublisher(ctx context.Context, config runtimeInterfaces.CloudEventsConfig, scope promutils.Scope) interfaces.Publisher {
```
The interfaces are the same now, but I do agree to have a new interface for cloudevent publisher. 
Yeah, we may extend `cloudevent.publisher` in the future.

katrogan commented on pkg/async/cloudevent/implementations/cloudevent_publisher.go:
```diff
@@ -0,0 +1,118 @@
+	cloudEventSource     = "https://github.com/flyteorg/flyteadmin"
+	cloudEventTypePrefix = "com.flyte.resource"
+	jsonSchemaURLKey     = "jsonschemaurl"
+	jsonSchemaURL        = "https://github.com/flyteorg/flyteidl/blob/cloudevent2/jsonschema/workflow_execution.json"
```
will you check this into master? can you hold off on merging this change until you do?

pingsutw replied:
yeah, we should merge this PR https://github.com/flyteorg/flyteidl/pull/270 first.  btw, will fix the test errors. cc @wild-endeavor
